Intro Level Tutorial

jcarnaxide opened the pull request:
*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
We have worked on an introductory level course for people to get familiar with Memray.

**Testing performed**
We ran `make docs-live` to view the new RST files, and to see everything is rendered properly.
We also ran the exercises via instructions specified and was able to get through the whole tutorial.

**Additional context**
It is worth noting, that I plan to update the instructions to include running the tutorial via codespaces, but we stuck with simple venv/pip for this first PR.

godlygeek commented on docs/index.rst:
```diff
@@ -33,3 +33,14 @@
    supported_environments
    changelog
    licenses
+
+.. toctree::
+   :hidden:
+   :caption: Hands-on Tutorial
+
+   intro
+   contents
+   exercise_1
+   exercise_2
+   exercise_3
+   additional_features
```
Let's actually yank this up higher - I think I'd like to create a new section called "Concepts" and drop `run` through `performance` into it, and then lift this section up to between `getting_started` and `run`. That way the sidebar will be ordered as:

- Overview
- Getting started
- Hands-on Tutorial
  - About these tutorials
  - Exercise 1 - ...
  - Exercise 2 - ...
  - Exercise 3 - ...
  - What to learn about next
- Concepts
   - run
   - python_allocators
   ...
```
I think that's a pretty nice organization, and should do a good job of highlighting these new docs. We should also make getting_started.html suggest that the next steps after you've generated your first flame graph would be to either try the tutorials or to learn more about `memray run`, with links to either the tutorial or skipping over it to the concepts section.

godlygeek reviewed (commented):
I haven't proofread any of the actual tutorials yet, but I've got a few comments on the organization and structure, to start with:

jcarnaxide commented:
@godlygeek I have addressed your comments for restructuring, as well as combining the introduction to the tutorials.
@statkute please have a look at the changes I have made.

I think we should still update the rst page for additional features inside the tutorials directory.

statkute commented:
> @godlygeek I have addressed your comments for restructuring, as well as combining the introduction to the tutorials. @statkute please have a look at the changes I have made.
> 
> I think we should still update the rst page for additional features inside the tutorials directory.

I've now updated a couple internal links that stopped working due to the new tutorials/ directory. Otherwise looks great-- thanks for the fixes while I was out @jcarnaxide !

I've also added a brief section that encourages to explore the `Concepts` section after finishing the hands-on tutorial (under additional features).

@godlygeek we should be good for another look whenever you have the time -- thank you!

godlygeek reviewed (approved):
I've made some edits, after which this LGTM.

I'm going to land this, so that it can be used by our Sprinters at PyCon US tomorrow.

@jcarnaxide and @statkute please take a look at the last commit on this PR if you get a chance, and open an issue or a new PR if you spot any mistakes I've introduced. Sorry for not giving you time to review before I landed it!
